ccsljava_execution/ccsljava_engine/plugins/org.eclipse.gemoc.execution.concurrent.ccsljavaengine/src/org/eclipse/gemoc/execution/concurrent/ccsljavaengine/dsa/executors/explorer/ExhaustiveConcurrentExecutionEngine.java

```diff
@@ -0,0 +1,543 @@
+import org.eclipse.gemoc.xdsmlframework.api.core.IExecutionEngine;
+import org.eclipse.gemoc.xdsmlframework.api.engine_addon.IEngineAddon;
+
+import fr.inria.diverse.k3.al.annotationprocessor.InitializeModel;
+import grph.Grph;
+
+/**
+ * Basic abstract implementation of the ExecutionEngine, independent from the technologies used for the solver, the
+ * executor and the feedback protocol. It can display the runtime execution information to its registered observers.
+ * 
+ * There are two phases of initializations for this entity:
+ * <ul>
+ * <li>the constructor sets the language-specific elements such as DomainSpecificEvents, Solver, EventExecutor,
+ * FeedbackPolicy</li>
+ * <li>the initialize method sets the model-specific elements such as Model and ModelLoader.</li>
+ * </ul>
+ * From the Model, we can derive :
+ * <ul>
+ * <li>the Model of Execution (using the DomainSpecificEvents)</li>
+ * <li>the Higher-order-transformation (TODO)) and the Solver Input (using the Model of Execution and the Solver Input
+ * Builder provided by the Solver).</li>
+ * </ul>
+ * 
+ * There are a few elements that could enter the ExecutionEngine, maybe after being reified as their own data
+ * structures:
+ * <ul>
+ * <li>{@link #scheduledSteps} is the FIFO of LogicalSteps. We use a FIFO because sometimes we may want to memorize
+ * steps in order to go back in the past and replay given steps. When the engine seeks to reach the next step of
+ * execution, it will first check if there is something in this FIFO. If there is not, then a new step is requested from
+ * the solver.</li>
+ * <li>{@link #schedulingTrace} is a map recording the order in which the steps of the solver have been used.</li>
+ * <li>{@link #scheduledEventsMap} records which step is linked to which list of MSEs. It is to be noticed that not all
+ * the MSEs were executed. Indeed, some may have been illegal at the time (with regards to the MoC) and thus discarded
+ * during the execution of this step.</li>
+ * <li>{@link #executionTrace} records the steps and the actual MSEs that were actually executed.</li>
+ * </ul>
+ * 
+ * @see IExecutionEngine
+ * 
+ * @author flatombe
+ * @author didier.vojtisek@inria.fr
+ * @author julien.deantoni@polytech.unice.fr
+ * @param <T>
+ * 
+ */
+public class ExhaustiveConcurrentExecutionEngine extends ConcurrentExecutionEngine{	
+	
+	public ExhaustiveConcurrentExecutionEngine(IConcurrentExecutionContext concurrentexecutionContext, ISolver s) throws CoreException 
+	{
+		super(concurrentexecutionContext,s);
+	}
+
+	@Override
+	public ILogicalStepDecider getLogicalStepDecider() {
+		return _logicalStepDecider;
+	}
+
+	@Override
+	public void changeLogicalStepDecider(ILogicalStepDecider newDecider) {
+		_logicalStepDecider = newDecider;
+	}
+
+	
+	
+	public StateSpace stateSpace = new StateSpace();
+	protected ArrayList<ControlAndRTDState> statesToExplore = new ArrayList<ControlAndRTDState>();
+	
+	/**
+	 * actually performs all the execution steps...
```

Comment:
This comment makes me wonder if imlpementing exhaustive exploration in the form of an ExecutionEngine makes sense. Shouldn't the engine only be responsible for a single concern which is to perform execution steps? And thus shouldn't exhaustive exploration be a tool that simply makes use of an engine in order to provide us with analysis results?

Note: I am far from being strongly opposed to having such an engine, I am just genuinely curious regarding what we should do, and of what are opinions on that.